plutus-core/plutus-ir/src/PlutusIR/Analysis/VarInfo.hs

```diff
@@ -0,0 +1,138 @@
+{-# LANGUAGE LambdaCase     #-}
+{-# LANGUAGE NamedFieldPuns #-}
+module PlutusIR.Analysis.VarInfo where
+
+import Control.Lens hiding (Strict)
+import PlutusCore.Core (funTyArgs)
+import PlutusCore.Core qualified as PLC
+import PlutusCore.Name qualified as PLC
+import PlutusIR.Core
+import Prettyprinter
+
+data VarsInfo tyname name = VarsInfo
+  { termVarInfoMap :: PLC.UniqueMap PLC.TermUnique (VarInfo tyname name)
+  , typeVarInfoMap :: PLC.UniqueMap PLC.TypeUnique (TyVarInfo tyname name)
+  }
+
+instance Semigroup (VarsInfo tyname name) where
+  (VarsInfo t ty) <> (VarsInfo t' ty') = VarsInfo (t <> t') (ty <> ty')
+
+instance Monoid (VarsInfo tyname name) where
+  mempty = VarsInfo mempty mempty
+
+lookupVarInfo ::
+  (PLC.HasUnique name PLC.TermUnique)
+  => name
+  -> VarsInfo tyname name
+  -> Maybe (VarInfo tyname name)
+lookupVarInfo name (VarsInfo vim _) = PLC.lookupName name vim
+
+lookupTyVarInfo ::
+  (PLC.HasUnique tyname PLC.TypeUnique)
+  => tyname
+  -> VarsInfo tyname name
+  -> Maybe (TyVarInfo tyname name)
+lookupTyVarInfo name (VarsInfo _ vim) = PLC.lookupName name vim
+
+data TyVarInfo tyname name =
+  NormalTyVar {}
+  | DatatypeTyVar { dtNumTyVars :: Int, dtConstructors :: [name] }
+
+data VarInfo tyname name =
+  NormalVar { varStrictness :: Strictness, varArity :: Maybe Arity }
+  | DatatypeConstructor { dcArity :: Arity , dcParentTyname :: tyname }
```

Comment:
Need some Haddock explaining what the "parent" of a constructor or a matcher means.

Reply:
If I get rid of the field names it will be even more obscure :(

Reply:
Added haddock anyway